src/goto-programs/abstract-interpretation/gcse.h

```diff
@@ -0,0 +1,167 @@
+  }
+  virtual bool ai_simplify(expr2tc &, const namespacet &) const override
+  {
+    return false;
+  }
+
+public:
+  bool merge(
+    const cse_domaint &b,
+    goto_programt::const_targett,
+    goto_programt::const_targett);
+  /// All expressions available
+  std::unordered_set<expr2tc, irep2_hash> available_expressions;
+
+protected:
+  /// Add non-primitive expression `e` (and its sub-expressions) into available_expressions.
+  void make_expression_available(const expr2tc &e);
+
+  /// Remove expression `e` (and everything that it depends on) from available_expressions
+  void havoc_expr(const expr2tc &e, const goto_programt::const_targett &);
+
+  /// Remove every expression from available_expressions that depends on symbol `sym`
+  void havoc_symbol(const irep_idt &sym);
+
+  // Helper function to check whether `src` depends on `taint`
+  bool should_remove_expr(const expr2tc &taint, const expr2tc &src) const;
+  // Helper function to check whether `src` depends on symbol `sym`
+  bool should_remove_expr(const irep_idt &sym, const expr2tc &src) const;
+
+public:
+  // TODO: clearly this shouldn't be here. The proper way is to create a new Abstract Interpreter
+  // that contains a points-to analysis
+  static std::shared_ptr<value_set_analysist> vsa;
+};
+
+#include <util/algorithms.h>
+/**
+ * @brief Global Common Subexpression Elimination algorithm
+ *
+ * Compute all common subexpression in a goto program. 
+ * For each common subexpression, a new intermediate variable 
+ * `__esbmc_cse_symbol$` is created and assigned to the common
+ * value.
+ *
+ * In ESBMC, the main advantage is for sequential dereferences/with
+ * statements. They generate lots of boilerplate to be dereferenced
+ * during symex.
+ *
+ * Before:
+ * X = a + b + c;
+ * Y = a + b + d;
+ *
+ * After:
+ * __ESBMC_cse_symbol$0 = a + b;
+ * X = __ESBMC_cse_symbol$0 + c;
+ * Y = __ESBMC_cse_symbol$0 + d;
+ */
+class goto_cse : public goto_functions_algorithm
+{
+public:
+  goto_cse(contextt &ns, std::shared_ptr<value_set_analysist> &vsa) : goto_functions_algorithm(true), context(ns), vsa(vsa)
+  {
+  }
+
+  virtual bool runOnProgram(goto_functionst &) override;
+  virtual bool
+  runOnFunction(std::pair<const dstring, goto_functiont> &F) override;
+
+  // TODO: we should have a method to convert an cse_symbol back to the original expr for CE.
+
+protected:
+  ait<cse_domaint> available_expressions;
+  contextt &context;
+  expr2tc obtain_max_sub_expr(const expr2tc &e, const cse_domaint &state) const;
+  void replace_max_sub_expr(
+    expr2tc &e,
+    const std::unordered_map<expr2tc, expr2tc, irep2_hash> &expr2symbol,
+    const goto_programt::const_targett &to,
+    std::unordered_set<expr2tc, irep2_hash> &matched_expressions) const;
+
+  symbolt
+  create_cse_symbol(const type2tc &t, const goto_programt::const_targett &to);
+
+private:
+  unsigned symbol_counter = 0;
+  const std::string prefix = "__ESBMC_cse_symbol";
+  std::shared_ptr<value_set_analysist> &vsa;
```

Comment:
Why is this a shared_ptr if only a reference to it is stored? Can't it be a reference to the value_set_analysist directly?